File: packages/api-models-ts/src/SummaryQuery.ts
/**
 * Request and response models for `/api/summary_query`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Schemas from "./Schemas.ts";

/**
 * Payload for a source summary similarity search.
 *
 * Exactly one of `q` (a free-text query) and `objID` (find sources similar to
 * that source's summary) must be given. `k` is the maximum number of sources
 * to return and must satisfy `1 <= k <= 100`; server default 5. `z_min` and
 * `z_max` bound the redshift of the returned sources and `classificationTypes`
 * restricts them to those classifications; omitting them applies no
 * restriction.
 *
 * @since 1.0.0
 * @category Models
 */
export interface SummaryQueryPost {
    readonly q?: string | undefined;
    readonly objID?: string | undefined;
    readonly k?: number | undefined;
    readonly z_min?: number | undefined;
    readonly z_max?: number | undefined;
    readonly classificationTypes?: ReadonlyArray<string> | undefined;
}

/**
 * One vector-store hit for a summary query (not a SkyPortal model).
 *
 * The shape is defined by the Pinecone client, not by SkyPortal: when `q` is
 * used the handler rebuilds each hit as exactly `id`, `score` and `metadata`,
 * but when `objID` is used it passes the raw `matches` of the Pinecone query
 * response straight through, so the remaining fields are Pinecone's
 * `ScoredVector` attributes (`values`, `sparseValues`) and may change with the
 * Pinecone SDK version rather than with SkyPortal. `metadata` holds whatever
 * SkyPortal indexed alongside the summary (`redshift`, `class`, ...), so it
 * stays free-form.
 *
 * @since 1.0.0
 * @category Models
 */
export const SummaryQueryMatch = Schemas.model(
    v.strictObject({
        id: v.string(),
        score: Schemas.NullishNumber,
        values: Schemas.nullish(v.array(v.number())),
        sparseValues: Schemas.nullish(Schemas.JsonObject),
        metadata: Schemas.nullish(Schemas.JsonObject),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type SummaryQueryMatch = v.InferOutput<typeof SummaryQueryMatch>;

/**
 * Results of a source summary similarity search.
 *
 * @since 1.0.0
 * @category Models
 */
export const SummaryQueryResults = Schemas.model(
    v.strictObject({
        query_results: Schemas.list(SummaryQueryMatch),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type SummaryQueryResults = v.InferOutput<typeof SummaryQueryResults>;

File: packages/api-models-py/skyportal_py_models/summary_query.py
"""Request and response models for ``/api/summary_query``."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class SummaryQueryPost(BaseModel):
    """Payload for a source summary similarity search."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    q: str | None = None
    obj_id: str | None = Field(alias="objID", default=None)
    k: int | None = None
    z_min: float | None = None
    z_max: float | None = None
    classification_types: list[str] | None = Field(
        alias="classificationTypes", default=None
    )


class SummaryQueryMatch(BaseModel):
    """One vector-store hit for a summary query (not a SkyPortal model).

    The shape is defined by the Pinecone client, not by SkyPortal: when
    ``q`` is used the handler rebuilds each hit as exactly ``id``,
    ``score`` and ``metadata``, but when ``obj_id`` is used it passes the
    raw ``matches`` of the Pinecone query response straight through, so
    the remaining fields are Pinecone's ``ScoredVector`` attributes
    (``values``, ``sparse_values``, serialized as ``sparseValues``) and
    may change with the Pinecone SDK version rather than with SkyPortal.
    ``metadata`` holds whatever SkyPortal indexed alongside the summary
    (``redshift``, ``class``, ...), so it stays free-form.
    """

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    id: str
    score: float | None = None
    values: list[float] | None = None
    sparse_values: dict[str, Any] | None = Field(alias="sparseValues", default=None)
    metadata: dict[str, Any] | None = None


class SummaryQueryResults(BaseModel):
    """Results of a source summary similarity search."""

    model_config = ConfigDict(extra="forbid")

    query_results: list[SummaryQueryMatch] = Field(default_factory=list)
